### DISCUSSION TITLE
Unknown crash

### DISCUSSION THREAD

**Discussion by nilesh-dryad** [jefftenney/LPTIM-Tick-U5#6] [General]
Hello Jeff,

You did wonderful job with tickless idle. It is working perfectly fine and saves a lot of power. I am running system with 160Mhz when it is not in idle mode and it also goes to 25Mhz during tickless mode.

I have following MACRO defined.
```
#define SUPPORT_VREG_RANGES_1_THROUGH_3
#define SUPPORT_MSI_AT_48MHZ
```

Sometime I am getting crash at two location randomly. I tried to find the cause but I couldn't find anything. I am attaching the image for the location where I am getting the crash.

![image](https://user-images.githubusercontent.com/80112251/202416454-ee521d8d-5dd9-4d76-9cbc-adbe3eb4cf24.png)

And line number 556 in below image.
![image](https://user-images.githubusercontent.com/80112251/202417247-cd228827-3915-4868-9cc6-d347e9e747f6.png)

It would be great help, if you have any idea.

**Comment by jefftenney** [comment]
Are these crashes occurring only while debugging?  If so, this is a known silicon bug on STM32L4+, and it is a silicon bug on the STM32U5 too.  See https://community.st.com/s/question/0D53W00001YaehVSAR/waking-from-stop-mode-causes-hard-fault-but-only-when-debugging.  See 2.2.16 in the [STM32L4+ errata](https://www.st.com/resource/en/errata_sheet/es0393-stm32l4rxxx-and-stm32l4sxxx-device-errata-stmicroelectronics.pdf).  The erratum is called "Wrong instruction fetches from flash memory upon wakeup from Sleep or Stop mode when debug in low-power mode is enabled".  As of this writing the STM32U5 errata sheet doesn't yet include this bug, but the bug is definitely in the silicon.  See [this commit](https://github.com/jefftenney/LPTIM-Tick-U5/commit/b5c6d51579ec04eb482c4fd27a97bd1d88b2f804) for the workaround I chose for the LPTIM-Tick-U5 demo program.

The silicon bug occurs only while debugging.  More precisely, the bug occurs only while `DBGMCU_CR_DBG_STOP` is set.  Be aware that this bit might remain set even after debugging ends depending on the debugger and on the way the debug session ended.  A power cycle is sure to clear `DBGMCU_CR_DBG_STOP`.

**Reply by nilesh-dryad** [reply]
Thanks @jefftenney for quick reply. The crash in the first image I mostly caught it while debugging but the crash in the second image was caught without debugging. In that case also I don't remember how I stopped debugging. 
But this is a good catch I will now remember to stop debugging and power cycle each time to debug the crash in the second image.
I will also try workaround for the crash in the first image. 

After looking at the workaround, I can see that MSI clock range is changed to 2 which means 16Mhz. Therefore I need to comment **SUPPORT_MSI_AT_48MHZ** macro. Even If I keep that marco uncommented, it will not affect (There will be few instruction more.) in case of MSI Range 2. 

For MSI RANGE 1 and 0, we definitely need that macro as per ICSCR1 register description.

**Reply by jefftenney** [reply]
The 2nd error (shown on line 556 above) is interesting.  It might be caused by the silicon bug, but maybe not.  Would be grateful if you could break into the debugger when that assert hits.  Check the values of `xFullTicksLeft` and `xExpectedIdleTime` as well as the LPTIM registers.  And anything else you see might be relevant.

**Reply by jefftenney** [reply]
If the 2nd error is "legitimate" (meaning, not caused by a silicon bug), then that error is caused by higher priority ISRs delaying the tick ISR for more than one tick duration (typically 1ms).  Do you have ISRs that can do that?  If you aren't sure, you could experiment with setting `configTICK_INTERRUPT_PRIORITY` to `configLIBRARY_MAX_SYSCALL_INTERRUPT_PRIORITY`.  If that solves the problem, then you do have ISRs that can delay the tick ISR for too long.  

For reference, see [this description](https://github.com/jefftenney/LPTIM-Tick-U5/blob/71a26d4f4ed8b58d23efb45a81bf80a2eb375274/Core/Src/lptimTick.c#L113-L119) of the requirement.

**Reply by nilesh-dryad** [reply]
Thanks @jefftenney I do confirm that first error was caused by the debugger because I didn't face after removing the debugger properly and power cycle the board.
For the second error which is coming rarely. I will look into interrupt priority and I will also print those two value when it comes into assert before going into assert.

**Reply by nilesh-dryad** [reply]
> The 2nd error (shown on line 556 above) is interesting. It might be caused by the silicon bug, but maybe not. Would be grateful if you could break into the debugger when that assert hits. Check the values of `xFullTicksLeft` and `xExpectedIdleTime` as well as the LPTIM registers. And anything else you see might be relevant.

xFullTicksLeft is **602** and xExpectedIdleTime is **110**.
Next time xFullTicksLeft is **1915**, xExpectedIdleTime is **209**.
Next time  xFullTicksLeft is **459**, xExpectedIdleTime is **196**

**Reply by jefftenney** [reply]
OK thank you.  That is helpful.

Can you try setting `configTICK_INTERRUPT_PRIORITY` to `configLIBRARY_MAX_SYSCALL_INTERRUPT_PRIORITY` now?

**Reply by nilesh-dryad** [reply]
Yes I will do that today and put it on over night test. One more thing I would like to inform you that I am using 10.4.6 FreeRTOS kernel.

I read below comment in lptimTick.c. Can this be the problem? 
`//      FreeRTOS version 10.4.0 or newer is recommended to ensure this function doesn't potentially return one
// OS tick *after* the intended time.`


Is it recommended to take changes from following commit?
https://github.com/FreeRTOS/FreeRTOS-Kernel/commit/195a351ec797ba434c0e00fbba1d1f6adb74e279

**Reply by jefftenney** [reply]
No need to worry about that comment you quoted because you are using 10.4.6. That version will work properly for you. No changes required.  No need to apply the commit you referenced.

**Reply by nilesh-dryad** [reply]
Hello
I checked my port.c in 10.4.6. It does not have the change that you committed in the freertos kernal repository. Unfortunately I didn't get time to look into that commit. But as it is not necessary, we will not apply it.

Yesterday we put two boards on overnight test. Unfortunately we don't use console logging but we use flash logging. 
One board was updated with this port.c change only (which you commit on freertos repos).
Another board was updated with port.c change and configTICK_INTERRUPT_PRIORITY change. 

Board with just port.c change crashed but board with both change did not crash.  We are looking for the reason in the flash logging. 
We are doing some more test to find the root cause. we will keep you posted.

**Reply by jefftenney** [reply]
You should be all set with the `configTICK_INTERRUPT_PRIORITY` change alone.

You do not need to apply any changes to port.c.  The commit to the FreeRTOS kernel (https://github.com/FreeRTOS/FreeRTOS-Kernel/commit/195a351ec797ba434c0e00fbba1d1f6adb74e279) makes no difference when you're using lptimTick.c.  Code in lptimTick.c replaces the affected functions in port.c.

**Reply by jefftenney** [reply]
Hi @nilesh-dryad, I just now committed 17e6318 which changes comments only.  Specifically, see the comment I added immediately above the `configASSERT()` statement.  Hopefully that will help the next person.

**Reply by nilesh-dryad** [reply]
@jefftenney Thank you for the comments. It helps. We have been testing with increased tick priority and since then we never got that asserts.

This means that we have some ISR which is blocking TICK ISR. We looked into our application and we have only have follwing ISR enabled in the NVIC. 

- EXTI4 -> 10 (Just a single line of code in the ISR to notify one Task using xTaskNotifyFromISR)

- IWDG -> 0 (This might be the problem, We used early wakeup ISR to make 5 minutes of watchdog timeout so we are doing some operations there. We have more than 15 FreeRTOS tasks.)

- TIM17 -> 15 (This we disabled in the main, therefore we won't get any interrupt)

All our FreeRTOS task has the same priority (tskIDLE_PRIORITY + 5) except one task.

- FreeRTOS timer task with (24). We are doing a lots of operations in Timer Callback continuously during FUOTA process and we found that we are getting assert only during FUOTA process. FUOTA process is too long in our application (7-8 hours). These timer callbacks also contains the code to write into external flash using SPI polling read write method. 

Therefore we believe that either IWDG ISR and these timer callbacks operation can be the problem of our asserts. As soon as we find something I will post it here.

**Reply by jefftenney** [reply]
The other thing that can delay the tick ISR is `configPOST_SLEEP_PROCESSING()`.  If you're using [this function](https://github.com/jefftenney/LPTIM-Tick-U5/blob/17e6318229d8a33690d8507ec1eaec88dccf6a36/Core/Src/ulp.c#L115-L153) unmodified as your `configPOST_SLEEP_PROCESSING()`, then you should be OK.  That function is very fast.

I think maybe I'll update my new comment to indicate the requirement that `configPOST_SLEEP_PROCESSING()` be short and fast.

**Reply by nilesh-dryad** [reply]
That function is as it is. SUPPORT_MSI_AT_48MHZ macro is enabled though.

**Reply by jefftenney** [reply]
In that case it's good to keep up your current investigation to determine why the tick ISR is being delayed for so long.  But you are safe to keep the fix you already have in place.  Increasing the priority of the tick ISR is a good clean fix.  It is suitable as a permanent fix.

EDIT: Forgot to mention that your timer callbacks do not delay the tick ISR.  Timer callbacks execute in the task context, not the ISR context.  So those timer callbacks are probably not related to the current issue.

**Reply by jefftenney** [reply]
Hi Nilesh, can you confirm that setting `configTICK_INTERRUPT_PRIORITY` to `configLIBRARY_MAX_SYSCALL_INTERRUPT_PRIORITY` resolved the assert issue (line 556 in your picture above)?

**Reply by nilesh-dryad** [reply]
@jefftenney We have testing this for so long now. Looks like we are getting line 556 assert more often now. 
```
#ifndef configTICK_INTERRUPT_PRIORITY
#define configTICK_INTERRUPT_PRIORITY configLIBRARY_MAX_SYSCALL_INTERRUPT_PRIORITY // default only; see above
#endif
```
`Apr 28 07:47:17 /0016c001f157f5de/u/statistics {"gwid":30174, "type":2, "CFSR":"00000000", "memfault":"00000000", "busfault":"00000000","stackoverflow":"", "taskwdt" : "", "r0":"00000000", "r1":"00000000", "r2":"00000000", "r3":"00000000", "r12":"00000000", "lr":"00000000", "pc":"00000000", "psr":"00000000", "pendInt0":"00000000", "pendInt1":"00000000", "pendInt2":"00000000", "pendInt3":"00000000","AssertPath":"_bg/lptimTick.c", "AssertLine":553, "FreeHeap":0, "ReqByte":0, "ErrHandler":0}
`
It seems that setting the tick priority is not resolving this crash.

**Reply by jefftenney** [reply]
Hi again Nilesh -

A few requests for you:

- Can you reply and attach your current copy of lptimTick.c?  It should have an assertion statement on line 553.
- What are all of the interrupts enabled in your application and their interrupt priorities?
- Can you post the content of your `configPOST_SLEEP_PROCESSING()` macro/function?
- Can you also attach your FreeRTOSConfig.h?

**Reply by nilesh-dryad** [reply]
Hello @jefftenney 
I have attached the interrupt Priorites set in the application. Please look at [attached file](https://github.com/jefftenney/LPTIM-Tick-U5/assets/80112251/d84b5899-97dc-47ee-8e07-06a2581afad0). 
LPTIM1 has configLIBRARY_MAX_SYSCALL_INTERRUPT_PRIORITY.

Please [click here](https://github.com/jefftenney/LPTIM-Tick-U5/files/15160125/lptimTick.txt) for **lptimTick.c**.
Please [click here ](https://github.com/jefftenney/LPTIM-Tick-U5/files/15160134/FreeRTOSConfig.txt) for **FreeRTOSConfig.h**.

And here is **configPOST_SLEEP_PROCESSING** function.

```
#define SUPPORT_VREG_RANGES_1_THROUGH_3
#define SUPPORT_MSI_AT_48MHZ

void vUlpPostSleepProcessing()
{
   if (SCB->SCR & SCB_SCR_SLEEPDEEP_Msk)
   {
      //      We may have been in deep sleep.  If we were, the RCC cleared several enable bits in the CR, and
      // it changed the selected system clock in CFGR.  Restore them now.  If we're restarting the PLL as
      // the CPU clock here, the CPU will not wait for it.  Instead, the CPU continues executing from the
      // wake-up clock (MSI in our case) until the PLL is stable and then the CPU starts using the PLL.

      #ifdef SUPPORT_VREG_RANGES_1_THROUGH_3
      {
         //      In the STM32U5, we must first restore the previous regulator range because it may be required
         // to support the previous clock speeds.  In deep sleep, the PWR module sets range 4 automatically.
         // Wait for the regulator (and booster if necessary) to be ready before restoring the clocks.
         //
         do
         {
            PWR->VOSR = pwrVosrSave;
         } while (PWR->VOSR != pwrVosrSave);
      }
      #endif

      #ifdef SUPPORT_MSI_AT_48MHZ
      {
         RCC->ICSCR1 = rccIcscr1Save;
      }
      #endif

      RCC->CR = rccCrSave;
      RCC->CFGR1 = rccCfgrSave;

      SCB->SCR &= ~SCB_SCR_SLEEPDEEP_Msk;

      //      This application bypasses the RTC shadow registers, so we don't need to clear the sync flag for
      // those registers.  They are always out of sync when coming out of deep sleep.
      //
      // RTC->ISR &= ~RTC_ISR_RSF;
   }
}
```

**Reply by jefftenney** [reply]
The only suspicious thing I saw was EXTI line 5.  It has a very high interrupt priority (zero, the highest).  What happens in that ISR?  Might it take a "long" time?  (I'm assuming that ISR does not make calls to the FreeRTOS API.)

Here's another idea.  As a diagnostic test, I would eliminate the use of STOP modes.  See if there is a connection there.  To eliminate STOP modes, you could comment out this line in `vUlpPreSleepProcessing()`:

`// SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;`

That might help you start to divide and conquer.

**Reply by nilesh-dryad** [reply]
@jefftenney Thank you for your reply. EXTI line 5 is added recently (just few days back) but it is just empty function because it is new feature that we are going to use in future. We can ignore that as of now but I can try disabling the STOP mode. I will keep you posted.

**Reply by nilesh-dryad** [reply]
@jefftenney I have an update. After removing STOP modes, none of my devices (total 6 devices) got assert at line 553. They are all running since more than 48 hours.

**Reply by jefftenney** [reply]
Does your firmware activate the ICACHE or DCACHE?  I'm thinking about erratum "2.2.11 ICACHE and DCACHE access might be corrupted after exiting Stop 2 and Stop 3 modes".

**Reply by nilesh-dryad** [reply]
Yes my firmware activates both but I am always using STOP1 mode. I have commented STOP2 part.
```
   int useDeepSleep = pdFALSE;
//   if (xDeepSleepForbiddenFlags == 0)
//   {
//      useDeepSleep = pdTRUE;
//      #define PWR_CR1_LPMS_STOP2 PWR_CR1_LPMS_1
//      MODIFY_REG(PWR->CR1, PWR_CR1_LPMS_Msk, PWR_CR1_LPMS_STOP2);
//   } else
//   if ((xDeepSleepForbiddenFlags & ~ulpPERIPHERALS_OK_IN_STOP1) == 0)
//   {
      useDeepSleep = pdTRUE;
	  #define PWR_CR1_LPMS_STOP1 PWR_CR1_LPMS_0
	  MODIFY_REG(PWR->CR1, PWR_CR1_LPMS_Msk, PWR_CR1_LPMS_STOP1);
//   }

   if (useDeepSleep)
   {
      rccCrSave = RCC->CR;
      rccCfgrSave = RCC->CFGR1;
      #ifdef SUPPORT_MSI_AT_48MHZ
      {
         rccIcscr1Save = RCC->ICSCR1;
      }
      #endif
      #ifdef SUPPORT_VREG_RANGES_1_THROUGH_3
      {
         pwrVosrSave = PWR->VOSR;
      }
      #endif

      SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
   }
```

**Reply by jefftenney** [reply]
In that case, the problem seems to be unrelated to ICACHE and DCACHE.  However, it might still be worth an experiment to turn them off and retest.  I think the root cause of the issue is probably unexpected hardware behavior.

Is it possible for you to build and test *without* SUPPORT_VREG_RANGES_1_THROUGH_3?  You would need to move the clock down to < 25 MHz, either permanently or at least any time `vUlpPreSleepProcessing()` might attempt STOP mode.  Here's what I'm thinking.  Your previous experiments tie the issue to the use of STOP mode.  My only theory to explain the issue is that something is delaying the tick ISR too much.  So maybe the wake-up-from-STOP-mode code is taking too long?  The wake-up code that supports VREG ranges 1 through 3 does look a little suspicious because it loops waiting (indefinitely) for the hardware to respond.

```
         do
         {
            PWR->VOSR = pwrVosrSave;
         } while (PWR->VOSR != pwrVosrSave);
```

For my part, I could make a branch of LPTIM-Tick-U5 that runs at 160 MHz and see if maybe the existing tests catch something.

**Comment by nilesh-dryad** [comment]
@jefftenney STM32U575 Errata sheet is also updated with this hardfault.
https://www.st.com/resource/en/errata_sheet/es0499-stm32u575xx-and-stm32u585xx-device-errata-stmicroelectronics.pdf

**Reply by jefftenney** [reply]
Thanks.  Erratum 2.2.18 in that document doesn't quite match 2.2.16 in the L4+ errata.  Wonder if its a separate issue.
